services/certificate/certificate-expiration.service.js

```diff
@@ -0,0 +1,110 @@
+import emojic from 'emojic'
+import Joi from 'joi'
+import { BaseService, pathParam, queryParam } from '../index.js'
+import trace from '../../core/base-service/trace.js'
+
+const queryParamSchema = Joi.object({
+  warningDays: Joi.number().min(0),
+  dangerDays: Joi.number().min(0),
+}).required()
+
+export default class CertificateExpiration extends BaseService {
+  static category = 'monitoring'
+
+  static route = {
+    base: 'certificate/expiration',
+    pattern: ':hostname',
+    queryParamSchema,
+  }
+
+  static openApi = {
+    '/certificate/expiration/{hostname}': {
+      get: {
+        summary: 'Certificate Expiration',
+        parameters: [
+          pathParam({ name: 'hostname', example: 'shields.io' }),
+          queryParam({
+            name: 'warningDays',
+            schema: { type: 'number' },
+            description:
+              'Warning (yellow) if less than this many days remain (default=30)',
+            example: '30',
+          }),
+          queryParam({
+            name: 'dangerDays',
+            schema: { type: 'number' },
+            description:
+              'Danger (red) if less than this many days remain (default=7)',
+            example: '7',
+          }),
+        ],
+      },
+    },
+  }
+
+  static _cacheLength = 0
+
+  async fetch({ hostname }) {
+    trace.logTrace('outbound', emojic.womanCook, 'trying', hostname)
+    let resp
+    try {
+      resp = await this._request({
+        url: `https://${hostname}/`,
+        options: {
+          followRedirect: false,
+        },
+      })
+      return { expiresStr: resp.res.req.socket.getPeerCertificate().valid_to }
+    } catch (err) {
+      // since got will throw on a redirect, we need to catch it here and possibly return the expiration date we found
+      const expiresStr =
+        err.response?.req?.socket?.getPeerCertificate()?.valid_to
```

Comment:
There's a couple of things here:

- We could end up here for multiple reasons - redirect, other non-success code (500, 404, whatever) or some other error.
- I don't think we need all the optional chaining operators. If we're assuming a response object has all these properties in the `try` block, we can also assume here that a response object has those properties in the catch block too.

I think my suggested implementation here would be something more like:

```
...
} catch (err) {
  if ((err instanceof ShieldsRuntimeError) && (Object.hasOwn(err, 'response'))) {
    return { expiresStr: err.response.res.req.socket.getPeerCertificate().valid_to }
  }
  throw err
}
```

Reply:
done